### PR TITLE
engine/exchange manager: Add support for custom exchanges

### PR DESCRIPTION
**Pull request by lrascao**
This PR allows applications that import GCT as a lib to declare custom exchanges that do not require full GCT integration.

- [X] New feature (non-breaking change which adds functionality)

- [X] go test ./... -race
- [X] golangci-lint run

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [X] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -105,6 +105,8 @@ func NewFromSettings(settings *Settings, flagSet map[string]bool) (*Engine, erro
 		return nil, fmt.Errorf("failed to create script manager. Err: %s", err)
 	}
 
+	b.ExchangeManager = SetupExchangeManager()
+
```
Comment (gloriousCode) [inline engine/engine.go:109]: This is already set in `Start()`. As it seems like you're using a different workflow to setup/load unsupported exchanges, the `ExchangeManager` is already exported and you should set it up from that context.

File: engine/exchange_manager.go [hunk 2]
```diff
@@ -45,10 +45,15 @@ var (
 	ErrExchangeFailedToLoad  = errors.New("exchange failed to load")
 )
 
+type CustomExchangeBuilder interface {
```
Comment (gloriousCode) [inline engine/exchange_manager.go:48]: Please add a comment

File: engine/exchange_manager.go [hunk 3]
```diff
@@ -183,6 +188,9 @@ func (m *ExchangeManager) NewExchangeByName(name string) (exchange.IBotExchange,
 	case "zb":
 		exch = new(zb.ZB)
 	default:
+		if m.Builder != nil {
+			return m.Builder.NewExchangeByName(nameLower)
+		}
```
Comment (gloriousCode) [inline engine/exchange_manager.go:193]: Please add test coverage

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Hi lrascao, just have some comments on the implementation

**Review by shazbert — COMMENTED** [review]
No issues from me (other than whats already been said), thanks for the PR.

File: engine/engine.go (hunk 1 above)
Comment (lrascao) [reply engine/engine.go:109]: That's the tricky bit, on my service (using this patch) i go:
```
engine.Bot, err = engine.NewFromSettings(&settings, flagSet)
...
engine.Bot.ExchangeManager.Builder = ExchangeBuilder{}
...
if err = engine.Bot.Start(); err != nil {
```
Main idea here is to have a functional `ExchangeManager` that can be tweaked around before `Start`.
Currently, (pre-patch)  `Start` creates and overwrites any existing `ExchangeManager` and copying the entire contents of `NewFromSettings` and `Start` seems a bit wasteful.

**Review by shazbert — APPROVED** [review]
Changes ACK. Also, this allows for the loading of the same base exchange multiple times which can make things quite unstable. Might be prudent doing a reflection check when adding the exchange but I will leave that up to you.

File: engine/engine.go (hunk 1 above)
Comment (gloriousCode) [reply engine/engine.go:109]: When reviewing yesterday it looked like you had left the original `bot.ExchangeManager = SetupExchangeManager()` in `Start()` and I didn't want the two of them there. So long as there is only one, I'm fine with it 👍

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes lrascao. As Shazbert pointed out, this does allow for loading multiple instances of the same exchange, do you still need https://github.com/thrasher-corp/gocryptotrader/pull/732 ?

**Review by thrasher- — APPROVED** [review]
tACK! Thanks @lrascao

**Comment by lrascao** [comment]
> tACK! Thanks for making those changes lrascao. As Shazbert pointed out, this does allow for loading multiple instances of the same exchange, do you still need #732 ?

👍🏻  @gloriousCode , you're right that the entire content of #732 is no longer needed, i'm splitting it into #756 and #757 to further facilitate the custom exchange use-case

**Pull request merged**